C error with anonymous functions

### Describe the bug

C error is produced when using anonymous functions in a structure. It used to compile in the past. Breaking changes make me sad :(

### Reproduction Steps

`v run client` on this repo:
https://github.com/islonely/shattlebip

### Expected Behavior

Expected code to compile and run.

### Current Behavior

C error being spit out.

```
================== C compilation error (from cc): ==============
cc: /tmp/v_501/client.01JF462PTRT83BEA41YTKEJHHB.tmp.c:13151:943: warning: braces around scalar initializer [-Wbraced-scalar-init]
cc: /tmp/v_501/client.01JF462PTRT83BEA41YTKEJHHB.tmp.c:13153:36: error: incompatible function pointer types initializing 'anon_fn_voidptr' (aka 'void (*)(void *)') with an expression of type 'void (main__App *)' (aka 'void (struct main__App *)') [-Wincompatible-function-pointer-types]
cc:  13153 |         _option_ok(&(anon_fn_voidptr[]) { main__frame }, (_option*)(&_t1), sizeof(anon_fn_voidptr));
cc:        |                                           ^~~~~~~~~~~
cc: /tmp/v_501/client.01JF462PTRT83BEA41YTKEJHHB.tmp.c:13155:52: error: incompatible function pointer types initializing 'anon_fn_term__ui__event_voidptr' (aka 'void (*)(struct term__ui__Event *, void *)') with an expression of type 'void (term__ui__Event *, main__App *)' (aka 'void (struct term__ui__Event *, struct main__App *)') [-Wincompatible-function-pointer-types]
cc:  13155 |         _option_ok(&(anon_fn_term__ui__event_voidptr[]) { main__event }, (_option*)(&_t2), sizeof(anon_fn_term__ui__event_voidptr));
cc:        |                                                           ^~~~~~~~~~~
cc: /tmp/v_501/client.01JF462PTRT83BEA41YTKEJHHB.tmp.c:13472:34: error: no member named '_v_do' in 'struct main__MenuItem'; did you mean '__v_do'?
cc:  13472 |                 main__Menu_selected(app->menu)._v_do();
cc:        |                                                ^~~~~
cc:        |                                                __v_do
cc: /tmp/v_501/client.01JF462PTRT83BEA41YTKEJHHB.tmp.c:2149:11: note: '__v_do' declared here
... (the original output was 15 lines long, and was truncated to 12 lines)
================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.8 2911f29.9e71e32

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.8 2911f29.9e71e32
OS: macos, macOS, 15.1.1, 24B91
Processor: 8 cpus, 64bit, little endian, Apple M3

getwd: /Users/adamoates/Documents/shattlebip
vexe: /Users/adamoates/v/v
vexe mtime: 2024-12-15 03:00:13

vroot: OK, value: /Users/adamoates/v
VMODULES: OK, value: /Users/adamoates/.vmodules
VTMP: OK, value: /tmp/v_501

Git version: git version 2.39.5 (Apple Git-154)
Git vroot status: weekly.2024.49-86-g9e71e324
.git/config present: true

CC version: Apple clang version 16.0.0 (clang-1600.0.26.4)
emcc version: N/A
thirdparty/tcc status: thirdparty-macos-arm64 713692d4
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzVlNTE3OGYxNDRmNTg1YWU0ZDUzYTgiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.pAfulurLNleNs60q3WwuuXucK8MqHtVapN76RQf7skk">Huly&reg;: <b>V_0.6-21608</b></a></sub>

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -2180,7 +2180,7 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 		// Skip "C."
 		name = util.no_dots(name[2..])
 	} else {
-		name = c_fn_name(name)
+		name = if is_selector_call { c_name(name) } else { c_fn_name(name) }
 	}
 	if g.pref.translated || g.file.is_translated || node.is_file_translated {
 		// For `@[c: 'P_TryMove'] fn p_trymove( ... `

--- a/vlib/v/tests/selector_call_with_reserved_c_name_test.v
+++ b/vlib/v/tests/selector_call_with_reserved_c_name_test.v
@@ -1,0 +1,21 @@
+struct Foo {
+}
+
+struct Bar {
+	do fn () bool = unsafe { nil }
+}
+
+fn (f Foo) get() Bar {
+	return Bar{
+		do: foobar
+	}
+}
+
+fn foobar() bool {
+	return true
+}
+
+fn test_main() {
+	t := Foo{}
+	assert t.get().do() == true
+}

